### PR TITLE
Add an index to dictionary of PMIx standard keys

### PR DESCRIPTION
**Pull request by rhc54**
Capture some ongoing work for footprint reduction by
assigning an index to the PMIx standard keys.

Signed-off-by: Ralph Castain <rhc@pmix.org>

### CODE REVIEW

**Comment by rhc54** [comment]
bot:lanl:retest

**Comment by rhc54** [comment]
@jjhursey Could you please take a gander at the PRRTE test container to see why `prte` is segfaulting? I cannot reproduce it. I've tried adding some output to the areas I changed, but I'm not seeing it here - so I guess it is failing somewhere else.

**Comment by rhc54** [comment]
@awlauria Any chance you could take a look at the PRRTE CI container and see where/how it is segfaulting? As I said above, I cannot replicate it and so we are kinda stuck.

**Comment by jjhursey** [comment]
I can try to reproduce this afternoon.

File: src/common/pmix_attributes.c [hunk 1]
```diff
@@ -58,13 +59,49 @@ static void atrkdes(pmix_attribute_trk_t *p)
...
+            p->type = dictionary[n].type;
+            p->description = pmix_argv_copy(dictionary[n].description);
+            pmix_hash_register_key(p->index, p);
+        }
+        /* include the internals */
+        for (n=0; UINT32_MAX != internals[n].index; n++) {
+            p = (pmix_regattr_input_t*)pmix_malloc(sizeof(pmix_regattr_input_t));
+            p->index = internals[n].index;
+            p->name = strdup(internals[n].name);
+            p->string = strdup(internals[n].string);
```
Comment (jjhursey) [inline src/common/pmix_attributes.c:100]: I've only been able to reproduce the crash with `--disable-debug` builds (if I `--enable-debug` then it does not crash). But the stack trace points to here:
```
(gdb) bt
#0  0x00007fffb5e2e6f4 in __strlen_power8 () from /lib64/libc.so.6
#1  0x00007fffb5e158e4 in strdup () from /lib64/libc.so.6
#2  0x00007fffb66304c0 in pmix_init_registered_attrs () from /workspace/exports/pmix/lib/libpmix.so.0
#3  0x00007fffb65d3844 in pmix_rte_init () from /workspace/exports/pmix/lib/libpmix.so.0
#4  0x00007fffb659fa1c in PMIx_server_init () from /workspace/exports/pmix/lib/libpmix.so.0
#5  0x00007fffb68abbb8 in pmix_server_init () at prted/pmix/pmix_server.c:810
#6  0x00007fffb6944bc8 in rte_init (argc=2, argv=0x10000ff0350) at ess_hnp_module.c:214
#7  0x00007fffb68822a0 in prte_init (pargc=0x7fffd1038db0, pargv=0x7fffd1038da8, flags=4 '\004') at runtime/prte_init.c:315
#8  0x0000000010007354 in main (argc=2, argv=0x7fffd103a4c8) at prte.c:515
```

Comment (jjhursey) [reply src/common/pmix_attributes.c:100]: I see. The problem is with this line in the first loop:
```
            p->description = pmix_argv_copy(dictionary[n].description);
```
and this part of the generated dictionary:
```
static pmix_regattr_input_t dictionary[] = {
    {.index = 0, .name = "PMIX_ATTR_UNDEF", .string = "pmix.undef", .type = PMIX_POINTER,
     .description = (char *[]){"NONE"}},
```
Notice that the description is not `NULL` terminated.

File: contrib/construct_dictionary.py [hunk 2]
```diff
@@ -315,10 +319,11 @@ def main():
         return 1
 
     # mark the end of the array
-    constants.write(""",
-    {.name = ""}
+    constants.write(""",\n
+    {.index = UINT32_MAX, .name = "", .string = "", .type = PMIX_POINTER, .description = (char *[]){"NONE"}}
```
Comment (jjhursey) [inline contrib/construct_dictionary.py:323]: The empty descriptions must be NULL termianted.
```suggestion
    {.index = UINT32_MAX, .name = "", .string = "", .type = PMIX_POINTER, .description = (char *[]){"NONE", NULL}}
```

File: src/common/pmix_attributes.c [hunk 3]
```diff
@@ -58,13 +59,49 @@ static void atrkdes(pmix_attribute_trk_t *p)
...
 
+/* define a couple of internal attributes */
+static pmix_regattr_input_t internals[] = {
+    {.index = PMIX_INDEX_BOUNDARY, .name = "PMIX_BFROPS_MODULE", .string = "pmix.bfrops.mod", .type = PMIX_STRING,
+        .description = (char *[]){"name of bfrops plugin in-use by a given nspace"}},
+
+    {.index = PMIX_INDEX_BOUNDARY+1, .name = "PMIX_PNET_SETUP_APP", .string = "pmix.pnet.setapp", .type = PMIX_BYTE_OBJECT,
+            .description = (char *[]){"blob containing info to be given to pnet framework on remote nodes", NULL}},
+
+    {.index = UINT32_MAX, .name = "", .string = "", .type = PMIX_POINTER, .description = (char *[]){"NONE"}}
```
Comment (jjhursey) [inline src/common/pmix_attributes.c:70]:
```suggestion
    {.index = UINT32_MAX, .name = "", .string = "", .type = PMIX_POINTER, .description = (char *[]){"NONE", NULL}}
```

File: src/common/pmix_attributes.c [hunk 4]
```diff
@@ -58,13 +59,49 @@ static void atrkdes(pmix_attribute_trk_t *p)
 }
 static PMIX_CLASS_INSTANCE(pmix_attribute_trk_t, pmix_list_item_t, atrkcon, atrkdes);
 
+/* define a couple of internal attributes */
+static pmix_regattr_input_t internals[] = {
+    {.index = PMIX_INDEX_BOUNDARY, .name = "PMIX_BFROPS_MODULE", .string = "pmix.bfrops.mod", .type = PMIX_STRING,
+        .description = (char *[]){"name of bfrops plugin in-use by a given nspace"}},
```
Comment (jjhursey) [inline src/common/pmix_attributes.c:65]:
```suggestion
        .description = (char *[]){"name of bfrops plugin in-use by a given nspace", NULL}},
```

File: contrib/construct_dictionary.py (hunk 2 above)
Comment (jjhursey) [reply contrib/construct_dictionary.py:323]: Similar change is needed for `PMIX_POINTER` on like 78

**Comment by jjhursey** [comment]
After making sure that the `.description` was always `NULL`-terminated I was able to get past the crash. I noted the ones I had to change in the PR, but you might want to double-check if there are other places.

**Comment by rhc54** [comment]
Aha!! Thanks Josh!

**Comment by rhc54** [comment]
bot:lanl:retest

**Comment by rhc54** [comment]
bot:ibm:prrte:retest

**Comment by rhc54** [comment]
bot:lanl:retest

Apr 22, 2022, 8:50:14 PM

**Comment by rhc54** [comment]
bot:ibm:prrte:retest

**Pull request merged**